### ISSUE TITLE
Support accommodation bookings

### ISSUE THREAD

**Issue by ghost** [puiterwijk/regcfp#157]
While we're here... for GUADEC 2017 we would like to include accommodation booking along with registration. This has been done in the distant past (GUADEC 2012 in Coruña for example) but hasn't been done since we started using regcfp, I think.

I guess broadly we need to write some kind of 'accommodation' module? I could work on this, if you can give me pointers on how you think it should be implemented.

I'll try and come up with a mockup for the actual registration page as well.

**Comment by ghost** [comment]
Thinking about this more, I think a generic 'registration-time purchases' system is the right thing to do. That could be used to sell T-shirts, pre-booked rooms, meal tickets, activities or anything else.

Here's an attempt at defining the requirements of such a setup:

 * At registration time, allow choosing an item to purchase from a list of options
 * Payment in full required at registration time (to prevent financial loss on prebooked items like accommodation)
 * Inventory tracking of the items that are being sold
 * Items should be marked as 'sold out' when no longer needed
 * Users should be able to view but not change their choice later on
 * Admins should be able to change the choice later on

And here's an example of what it could look like in the config.json file:

```
      "sales": {
        "t_shirts": {
          "display_name": "T-shirt",
          "short_display_name": "T-shirt",
          "type": "select",
          "message": "This will be kept private",
          "required": false,
          "private": true,
          "options": {
            "XS": {
                "limit": 2,
                "cost": 20
            },
            "S": {
                "limit": 4,
                "cost": 20
            },
            "M": {
                "limit": 15,
                "cost": 20
            },
            "L": {
                "limit": 5,
                "cost": 20
            },
            "XL": {
                "limit": -1,
                "cost": 20
            },
            "baby": {
                "limit": 0,
                "cost": 20
            }
          }
        },
        "rooms": {
          "display_name": "Accommodation booking",
          "short_display_name": "Accommodation",
          "type": "select",
          "required": false,
          "private": true,
          "options": {
            "1 night": {
                "limit": 2,
                "cost": 20
            },
            "2 nights": {
                "limit": 4,
                "cost": 40
            }
          }
        }
      },
```

Does this sound like a sensible approach?

**Comment by benzea** [comment]
Just a note, we had something similar for 2016 in Karlsruhe. The workshops were a normal "drop down" field and with a limitation on how many people can selected. With regard to payment it was only a client side calculation though.

I think the approach outlined here should work in general. Main issue is probably with regard to payment. Changes are at least required if you want to enforce advanced payment for these items but also allow on-site payments for simple registrations as is usual for GUADEC.

**Comment by ghost** [comment]
Yes, we'll want to enforce advanced payment for the rooms. I'll look at making that possible.
Is the code for the workshop sales available somewhere? I can't find it in any branches of this repo

**Comment by benzea** [comment]
Hacks upon hacks. So it was client side, the server side config is just the field:

        "workshop-doc" : {
          "display_name": "Workshop registration",
          "type": "documentation",
          "html": "If you like you can join one of the following <a href=\"https://2016.guadec.org/workshops/\" target=\"_blank\">workshops on 11. August (before the core days)</a>."
        },
        "workshop": {
          "display_name" : "",
          "short_display_name": "Workshop",
          "type": "select",
          "required": false,
          "options":
            [ "Not participating in a workshop",
              "Packaging your application using Flatpak (formerly xdg-app)",
              "GStreamer",
              "Make your first contribution to GNOME",
              "Contribute to an OpenSource project" ],
          "limits" : [-1, 16, 16, 16, 15 ],
          "message" : "Attendee number is limited. We may contact you with workshop instructions, for payment purposes, and to confirm your attendence.",
          "private": true
        },


Then there are parts in gnome-reg.js, style-gnome.css and also register.hbs (see div with id "workshop-payments"). The only added support in main regcfp is the number limitation (commit c6a1788b322f8c3d38dfb1e8ebae1683d2033894 "Add support to limit number of responses in select field"). The other modifications are in commit 973f9066e80fdfe9f54a2fdac19b5bd95bbe929e "gnome: Add workshop payment into normal registration." (probably with further changes later).

**Comment by ghost** [comment]
Fixed by #167
